### ISSUE TITLE
Email Address allowed not conforming to RFC with FORWARD SLASH / and QUESTION MARK ? inclusion

### ISSUE THREAD

**Issue by jimimaseye** [hmailserver/hmailserver#173]
MAIL FROM and RCPT TO email addresses incorrectly being refused.

RFC2822 says forward slash and question marks are valid characters: Page 14 : https://www.ietf.org/rfc/rfc2822.txt:

```
addr-spec       =       local-part "@" domain
local-part      =       dot-atom / quoted-string / obs-local-part
domain          =       dot-atom / domain-literal / obs-domain
dot-atom        =       [CFWS] dot-atom-text [CFWS]
dot-atom-text   =       1*atext *("." 1*atext)
atext           =       ALPHA / DIGIT / ; Any character except controls,
                        "!" / "#" /     ;  SP, and specials.
                        "$" / "%" /     ;  Used for atoms
                        "&" / "'" /
                        "*" / "+" /
                        "-" / "/" /
                        "=" / "?" /
                        "^" / "_" /
                        "`" / "{" /
                        "|" / "}" /
                        "~"
specials        =       "(" / ")" /     ; Special characters used in
                        "<" / ">" /     ;  other parts of the syntax
                        "[" / "]" /
                        ":" / ";" /
                        "@" / "\" /
                        "," / "." /
```
FURTHER TESTS show the following characters being incorrectly denied in Hmailsever (5.6.5 - 2360) : 
**?
/
*
|**
These characters are listed as allowed in the ATEXT values (which form 'dot-atom') and are not in the 'specials' exclusions.

eg, **geoff/home@domain.com**  should be allowed
 **geoff?home@domain.com**  should be allowed

And in theory according to the wording: 
 **geoff@domain?num1.com**  should be allowed
 **geoff@domain/num1.com**  should be allowed

Websites sometimes use the forward slash in their email address which should be valid.

dnswl.org (public whitelist) uses such form in their validation/registration confirmation and is being rejected.

**hostmaster+public/www-data@dnswl.org**  should be allowed

https://www.hmailserver.com/forum/viewtopic.php?f=7&t=29812

(As with QUOTED locals, mailboxes requiring to utilise some of these characters in their account name that otherwise break windows file system could operate the ALIAS method to point to a real account.  However, emails addresses with these characters should be allowed for SEND and RECEIPT addresses to conform to RFC.)

(p.s  Tested "**test/me|pipe@yahoo.com**" (including the pipe symbol and forward slash) with gmail to see if they are fully compliant and it was accepted as expected.  Although "geoff@home?num1.com"  and "geoff@home*num.com" was not.)

**Comment by jimimaseye** [comment]
Users have reverted back a version to restore the functionality of allowing these addresses as its a real problem.

**Comment by martinknafve** [comment]
I've done some tweaks now. When setting up an account, it needs to confirm to a more stricter expression:

^(("[^<>@\\/\?\*|]+")|([^<> @\\/"\?\*|]+))@[^<>"\\/@\?\*| ]+\.[^<>"\\/@\?\*| ]+$

But as sender and recipient address, the following more allowing variant is used:

^(("[^<>@\\]+")|([^<> @\\"]+))@[^<>"\\/@\?\*| ]+\.[^<>"\\/@\?\*| ]+$

You can use for example http://regexr.com/ to test whether addresses matches these.

So hostmaster+public/www-data@dnswl.org for example will be allowed as sender/recipient address but not as an account address.

**Comment by jimimaseye** [comment]
MArtin a quick test of the valid RECIPIENT address shows an error.

You state:  ^(("[^<>@\]+")|([^<> @\"]+))@[^<>"\/@\?*| ]+.[^<>"\/@\?*| ]+$

but it needs to be 
^(("[^<>@\\]+")|([^<> @\\"]+))@[^<>"\/@\?*| ]+\.[^<>"\/@\?*| ]+$

(the BACKSLASH, and DOT needs to be escaped to be compatible with regex as they otherwise have a meaning rather than literal characters)

I would think you more restricted regex needs the same:
^(("[^<>@\/\?*|]+")|([^<> @\/"\?*|]+))@[^<>"\/@\?*| ]+\.[^<>"\/@\?*| ]+$

(I did do testing to prove first)

**Comment by jimimaseye** [comment]
I noted in your code that you use double backslash for 'escaping' (but 
strangely not always.  I cant say I understand why).  You would of 
course need to add the extra backslashes to the tested regex strings 
above if its necessary for double backslash.

Reminder of new tested regexpression: (without the double slashing) 
modifying your previous suggestion that had faults:

RECIPIENT valid address:

|^(("[^<>@\\]+")|([^<> @\\"]+))@[^<>"\/@\?*| \.\[]+(|\.[^<>"\/@\?*| 
\.\[]+(|\.[^<>"\/@\?*| \.\[]+(|\.[^<>"\/@\?*| \.\[]+(|\.[^<>"\/@\?*| 
\.\[]+))))$|

and

ACCOUNT mailbox (with your more restricted 'local-part'):
|^(("[^<>@\/\?*|\\]+")|([^<> @\/"\?*|\\]+))@[^<>"\/@\?*| 
\.\[]+(|\.[^<>"\/@\?*| \.\[]+(|\.[^<>"\/@\?*| \.\[]+(|\.[^<>"\/@\?*| 
\.\[]+(|\.[^<>"\/@\?*| \.\[]+))))$
|

||
* prevents double dots (user@domain..com)
* allows single domain (user@localhost)
* allow UP TO 5 level domain  (user@one.two.three.four.five)
* doesnt allow square brackets in domain (user@[12.34.56.78] )
* does allow ip address (user@12.34.56.78)
* prevents disallowed backslash in local-part

Hope you incorporate it (i spent hours working this out  ;-)  )

Jim

**Comment by martinknafve** [comment]
Thanks dude. Looking at the visualization of these regexes over at https://regexper.com still hints about an error though. The following address is allowed: test@test].com, but the following is not: test@test[.com. Also they currently allow { and }.

The original regexes were not 100% correct or complete, but I chose to go with that approach to make the regex a bit easier to read (to make them maintainable). As you noted yourself earlier, a 100% proper regex is impossible to read (kilobytes of regex). If the regexes gets too complex, they can't be understood easily and then painful to bugfix.

Reading wikipedia, The domain part of the email has to match:

> The domain name part of an email address has to conform to strict guidelines: it must match the requirements for a hostname, a list of dot-separated DNS labels, each label being limited to a length of 63 characters and consisting of:[7]
> uppercase and lowercase Latin letters A to Z and a to z;
> digits 0 to 9, provided that top-level domain names are not all-numeric;
> hyphen -, provided that it is not the first or last character.

Based on that, what o you think about the following regex for the domain-part of the user address:

Replacing the domain part in your last example with:

([a-zA-Z0-9-]+\.?){1,127}

For example then:
^(("[^<>@\\]+")|([^<> @\\"]+))@([a-zA-Z0-9-]+\.?){1,127}$

In other words, any character (a-z, any digit, and hyphen). In theory, a host name can have up to 127 sub domains so I think it makes reasonable to support that. Limiting it to 5 would prevent valid addresses such as example@internal.adm.div.stuff.co.uk (which is of course long but would not suprise me).

This variant supports the things you mentioned above, but also prevent rubbish such as:

user@fdsafdsa{}¤!=%).com

**Comment by jimimaseye** [comment]
I agree with the limit on 5 domains potentially being a problem.  At the 
time I simply couldnt remember how to do the {1,127} repitition test.  A 
much better offering.

Regarding my former expression allowing {curly brackets}:  well 
according to the RFC 2822 (quoted at the start of this discussion, 
opening post) they should be allowed.  How correct the WIKI page is that 
you refer to then I dont know.

Anyway, I did a test with your latest proposed expression:

([a-zA-Z0-9-]+.?){1,127}

but it still fails to stop strings such as

*user@random[squarebracket.com*

So I assumed you meant the DOT (before the question mark) literally as a 
character. So escaping it as: ([a-zA-Z0-9-]+\.?){1,127}it then fails to stop

*user@endswithdot.**
**user@domain.endswithdot.*

So......

Accepting your proposed string of characters([a-zA-Z0-9-]+\.?)I propose 
the following domain part:

^@[^-\.][a-zA-Z0-9-]+(?<!-)(|\.[^-\.][a-zA-Z0-9-]+(?<!-)){1,126}$

Note: this offer INCLUDES the escaped dot for literal character).

This allows you chosen string to exist once without an ending dot PLUS 
subsequent repetitions (up to 126 times) that must begin with a dot ( 
which would total 127 domains allowed.)

allowing:

*user**@domain**
**user@domain1.**domain2 ....etc to 127 domains**
*
disallows

*user@-startswithdash.com*  (as per your wiki page)
*user@endswithdash-.com*  (as per your wiki page)
*user@endswithdash-*  (as per your wiki page)
*user@-startdash.com*  (as per your wiki page)
*user@**endswithdot.*
*user@d**ouble..dot*

I checked it with https://regex101.com

HOWEVER, there are different flavours of regex and the 'negative 
lookbehind' (?<!  ) works with PHP/PCRE  but does not work with 
Javascript regex.  So I would suggest you could try it in your c++ code 
to see if it complains with this expression or not.

If you are more limited to the javascript version that doesnt allow 
'lookbehind' then I suggest:

^@[^-\.][a-zA-Z0-9-]+(|\.[^-\.][a-zA-Z0-9-]+){1,126}$

this at least stops

disallows:

*user@-startswithdash.com
**user@domain.-startswith**dash**
user@.startswithdot.com
*
but doesnt stop domain parts ending with a dash  (therefore 
*user@**endfswithdash-.com* would be allowed).


What do you think?  (The first one is very close to 'sufficiently 
perfect', allowing the confusion over RFC/wiki clash).

**Comment by jimimaseye** [comment]
GOT IT!!!

Fully working with all flavours of regexp:

`^(("[^<>@\\]+")|([^<> @\\"]+))@[^-\.][a-zA-Z0-9-]+[^\-\.](|\.[^-\.][a-zA-Z0-9-]+[^\-\.]){1,126}$`

(Note this is escaped)

Stops:
**user@-startswithdash.com
user@-startwithdash
user@domain.-startswithdash
user@domain.lastcharacterdash-**

**user@.startswithdot.com
user@endswithdot.
user@double..dot**

Basically stopping all permutations of incorrect placements of DASH and DOT, whilst allowing only the characters [a-zA-Z0-9-] in all positions.

**Comment by jimimaseye** [comment]
Just to follow on from my last wonder-update, and re-reading your last post, you mention 'maintainable' and readable as important.

So I will explain the expression (this is split with spaces and **bold**ing for readability)

^  **(("[^<>@\\\\]+")|([^<> @\\\\"]+))**     @      **((?!-|\\.)[a-zA-Z0-9-]{0,62}[^\\-\\.])**    (|\\.   **(?!-|\\.)[a-zA-Z0-9-]{0,62}[^\\-\\.])**    {1,126}$



The first part you know:

**(("[^<>@\\\\]+")|([^<> @\\\\"]+))**

This is the local part allowing "quoted" or not quoted

@   = must be @ to separate local from domain

then

**((?!-|\\.)[a-zA-Z0-9-]{0,62}[^\\-\\.])**

1st domain part: must not start or end with dot or DASH, contains valid characters and is no more than 63 characters.

then

(|\\.

either null (there is no more to follow) OR a DOT to signify another domain to follow

then

**(?!-|\\.)[a-zA-Z0-9-]{0,62}[^\\-\\.])**

another DOMAIN spec. As the first, must not start or end with dot or dash (but forms part of the _null OR ..._ nest)

{1,126}$

allow the second [dot]domain to be repeated up to 126 times.

Hopefully this shows the simplicity.

subsequent domains must start with DOT (as the separator)

They key point is that there are 2x domain spec sections - one for the first occurrence immediately after the @ (without any dots preceeding or following) and then subsequent ones with an obligatory DOT preceeding it.

**Comment by martinknafve** [comment]
I admire your patience with this. :) Unfortuantely there are still cases which the latest regex fails to find. Latest regex:

    ^(("[^<>@\\]+")|([^<> @\\"]+))@(\[([0-9]{1,3}\.){3}[0-9]{1,3}\]|(?=.{1,255}$)((?!-|\.)[a-zA-Z0-9-]{0,62}[^\-\.])(|\.(?!-|\.)[a-zA-Z0-9-]{0,62}[^\-\.]){1,126})$

This regex accepts the following invalid addresses:

* test@example.Ö.com (should not be valid since Ö can't be part of domain name)
* test@example/.com (should not be valid since/ should not be part of domain name)

Specifically, it seems like it's the following sub-expression which causes the issue:

    [^\-\.]

We say that the domain may not end with dot or dash, but with this expression it may end with ANY character except dot or dash, including ¤, \, % or whatever.

If I replace this sub section with [a-zA-Z0-9], then it seems to work properly. Then rather than blacklisting dash or dot, we specify that only a-Z0-9 are allowed as the last character. In other words the latest version is then:

    ^(("[^<>@\\]+")|([^<> @\\"]+))@(\[([0-9]{1,3}\.){3}[0-9]{1,3}\]|(?=.{1,255}$)((?!-|\.)[a-zA-Z0-9-]{0,62}[a-zA-Z0-9])(|\.(?!-|\.)[a-zA-Z0-9-]{0,62}[a-zA-Z0-9]){1,126})$

Just FYI: I have added the examples you given to the automated test suite for hMailServer and this expression seems to behave properly for those.

![image](https://cloud.githubusercontent.com/assets/71842/15482517/695c3756-2130-11e6-83b3-66f092e6a6aa.png)

**Comment by jimimaseye** [comment]
Thanks for the gratitude for my persistence.  I have been called tenacious my employers in the past - I just call it stubborn obsessiveness.  (Hey, if its going to be done, then why not do it right, ai.)

I agree (and can see why) that final [\-\.] caused the problem - in fairness I didnt test for dodgy swedish characters ;-D. Seriously, yes, I now see why you exchange of that for [a-zA-Z0-9] makes it work correctly.

Good call, I think you have the perfect solution there.

BUT....do you want to incorporate the check for IPV6 literal addressing too (as well as the ipv4 format already catered for)? If so, here it is:

`^(("[^<>@\\]+")|([^<> @\\"]+))@(\[([0-9]{1,3}\.){3}[0-9]{1,3}\]|\[(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\]|(?=.{1,255}$)((?!-|\.)[a-zA-Z0-9-]{0,62}[a-zA-Z0-9])(|\.(?!-|\.)[a-zA-Z0-9-]{0,62}[a-zA-Z0-9]){1,126})$`

Allows

test@[1762:0:0:0:0:B03:1:AF18]
test@[1762:0:0:0:0:b03:1:af18]
(mix of upper and lowercase allowed)

 Or should we leave that for now?  :-)

**Comment by jimimaseye** [comment]
Another to cover part of RFC rules....

RFC says that domains must not be all numeric (there must be at least one alpha character)
eg,   12345  is not allowed (so  @12345.com or mail.1234.com is disallowed)

also, LOCAL-PART must not start with [.]DOT, end with [.]DOT or have a [..]DOUBLE DOT in an unquoted local-part.

So this regex is of the above (including digits shortened to \d and allowing:

allows:

ipv4 literals "@[123.45.678.9]" 
ipv6 literals "@[IPv6:2001:db8:85a3:8d3:1319:8a2e:370:7348]"   (https://tools.ietf.org/html/rfc5321#section-4.1.3)

disallows:

**user@1234
user@1234.com
user@domain.1234

.startswithdot@domain.com
endswithdot.@domain.com
double..dot@domain.com**

`^("[^<>@\\]+"|(?!\.|.*\.{2}|.*\.@)[^<> @\\"]+)@(\[([0-9]{1,3}\.){3}[0-9]{1,3}\]|\[IPv6:(?:[A-Fa-f0-9]{1,4}:){7}[A-Fa-f0-9]{1,4}\]|(?=.{1,255}$)((?!-|\.|\d+($|\.))[a-zA-Z0-9-]{0,62}[a-zA-Z0-9])(|\.(?!-|\.|\d+($|\.))[a-zA-Z0-9-]{0,62}[a-zA-Z0-9]){1,126})$`

(code is escaped)

Tested with regexper.com: https://regexper.com/#^%28%22[^%3C%3E%40\\]%2B%22|%28%3F!\.|.*\.{2}|.*\.%40%29[^%3C%3E%20%40\\%22]%2B%29%40%28\[%28[0-9]{1%2C3}\.%29{3}[0-9]{1%2C3}\]|\[IPv6%3A%28%3F%3A[A-Fa-f0-9]{1%2C4}%3A%29{7}[A-Fa-f0-9]{1%2C4}\]|%28%3F%3D.{1%2C255}%24%29%28%28%3F!-|\.|\d%2B%28%24|\.%29%29[a-zA-Z0-9-]{0%2C62}[a-zA-Z0-9]%29%28|\.%28%3F!-|\.|\d%2B%28%24|\.%29%29[a-zA-Z0-9-]{0%2C62}[a-zA-Z0-9]%29{1%2C126}%29%24

It's going well.  :-)

**Comment by martinknafve** [comment]
I'll go with the slightly simpler option we derived to earlier (the second last, rather than the last). This does not have to be 100% correct - If a user enters an invalid recipient address, he will have a bounce message stating that the address is invalid. Of course it's better to detect this as early as possible, but I'd rather keep the regex slightly easier, than trying to account for all different variants.

**Comment by jimimaseye** [comment]
Oh.

I thought the correct coding for the local-part formats (leading DOTS, double DOTS etc) was quite important.  Most email servers I tested stops these at delivery source.  After all, isnt that what started this bug in the first place?  (quotes and slashes not allowed in local-part)
